ENH: Add warning when setting into nonexistent attribute

deniederhut opened the pull request:
- [ ] closes #7175
 - [ ] closes #5904
 - [ ] tests added / passed
 - [ ] passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [ ] whatsnew entry

deniederhut committed:
ENH: adds warning when setting list-like into attribute

Adds warning in generic setattr logical branch for when attribute does
not exist and user is supplying a list-like object. Warning states that
Series cannot be assigned into nonexistent columns, and includes a link
to stable documentation. Closes #7175.

deniederhut committed:
TST: Adds tests for warning behavior on set on nonexistent attr

deniederhut committed:
DOC: replaces silent error for setattr on nonexistent name with UserWarning

deniederhut committed:
CLN: juggles doc url to fit within 80 char per line

pep8speaks commented:
Hello @deniederhut! Thanks for submitting the PR.

 - In the file [`pandas/core/generic.py`](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py), following are the PEP8 issues :

> [Line 351:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L351): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 364:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L364): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 581:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L581): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 864:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L864): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1758:17](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L1758): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1867:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L1867): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1875:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L1875): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 2636:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L2636): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 2845:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L2845): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 3289:13](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L3289): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 3304:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L3304): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 3309:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L3309): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 3314:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L3314): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 3327:17](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L3327): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 5369:25](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L5369): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 5409:29](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L5409): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 6253:9](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/core/generic.py#L6253): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def

 - In the file [`pandas/tests/dtypes/test_generic.py`](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/tests/dtypes/test_generic.py), following are the PEP8 issues :

> [Line 45:31](https://github.com/pandas-dev/pandas/blob/eb4668694a2a4a0e9f2ab7d08b5e211c1dea1029/pandas/tests/dtypes/test_generic.py#L45): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

jreback commented on doc/source/indexing.rst:
```diff
@@ -226,7 +226,7 @@ as an attribute:
 
 You can use attribute access to modify an existing element of a Series or column of a DataFrame, but be careful;
-if you try to use attribute access to create a new column, it fails silently, creating a new attribute rather than a
+if you try to use attribute access to create a new column, it issues a `UserWarning` and creates a new attribute rather than a
```
use double-backticks around ``UserWarning``

jreback commented on pandas/core/generic.py:
```diff
@@ -3237,6 +3237,11 @@ def __setattr__(self, name, value):
+                if (self.ndim > 1) and (is_list_like(value)):
+                    warnings.warn("Pandas doesn't allow Series to be assigned "
+                                  "into nonexistent columns - see "
+                                  "https://pandas.pydata.org/pandas-docs/"
```
you need a stacklevel=2 (or maybe higher here)

deniederhut replied:
This has been fixed -- not sure why GH isn't picking it up

jreback replied:
use ``if isinstance(self, ABCDataFrame) and is_list_like(value):``

jreback replied:
this change here

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,8 @@ def test_abc_types(self):
+        with catch_warnings(record=True) as w:
+            self.series.not_an_index = [1, 2]
+            assert len(w) == 0 # fail if false warning on Series
+        with pytest.warns(UserWarning):
```
use ``tm.assert_raises_warning``

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,8 @@ def test_abc_types(self):
         assert isinstance(self.sparse_array, gt.ABCSparseArray)
         assert isinstance(self.categorical, gt.ABCCategorical)
         assert isinstance(pd.Period('2012', freq='A-DEC'), gt.ABCPeriod)
+        with catch_warnings(record=True) as w:
```
what are you testing here?

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,8 @@ def test_abc_types(self):
         assert isinstance(pd.Period('2012', freq='A-DEC'), gt.ABCPeriod)
+        with catch_warnings(record=True) as w:
+            self.series.not_an_index = [1, 2]
+            assert len(w) == 0 # fail if false warning on Series
```
move this to a new test add the issue number as a comment

jreback commented:
Replicate this test from the original issue (and assert each of these things).

```
d = {'one' : pd.Series([1., 2., 3.], index=['a', 'b', 'c']),
     'two' : pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd'])}
df = pd.DataFrame(d)

# this adds a column to df
df['three'] = df.two + 1

# this successfully modifies an existing column
df.one += 1

# this seems to work but does not add a column to df
df.four = df.two + 2

print df.three
print df.four
print df
```

I would be ok with this raising
```
# this seems to work but does not add a column to df
df.four = df.two + 2
```

jreback commented:
also would love to warn on this: https://github.com/pandas-dev/pandas/issues/5904 as well.

deniederhut commented:
Wilco. Still in `test_generic.py`?

jreback commented:
> Wilco. Still in test_generic.py?

yeah that seems reasonable

deniederhut committed:
BUG: replaces single backticks with double backticks for rst

jreback commented on doc/source/indexing.rst:
````diff
@@ -226,7 +226,7 @@ as an attribute:
 
 You can use attribute access to modify an existing element of a Series or column of a DataFrame, but be careful;
-if you try to use attribute access to create a new column, it fails silently, creating a new attribute rather than a
+if you try to use attribute access to create a new column, it issues a ```UserWarning`` and creates a new attribute rather than a
````
maybe add a: starting in 0.21.0

deniederhut committed:
BUG: adds stacklevel to warning call

deniederhut committed:
DOC: adds 0.21.0 for appearance of setting warnings

deniederhut committed:
ENH: Add warning when colname collides with methods

Current behavior does not allow attribute-like access when the column
name of a DataFrame matches the name of a method. This commit adds an
explicit warning about this behavior. Closes #5904.

deniederhut committed:
TST: Adds testing around warnings for setting attributes

jreback commented on pandas/core/generic.py:
```diff
@@ -1787,6 +1787,10 @@ def _slice(self, slobj, axis=0, kind=None):
     def _set_item(self, key, value):
+        if callable(getattr(self, key, None)):
+            warnings.warn("Pandas doesn't allow attribute-like access to "
+                          "columns whose names collide with methods",
```
i would say that this is a collision

we actually DO allow it (maybe should raise but let's start with a warning)

deniederhut replied:
The warning says that attribute-like access is not allowed. E.g.

```python
df = pd.DataFrame({'a': [1, 2]})
df['sum'] = [3, 4]
print(df.sum)
```

```
<bound method DataFrame.sum of    a  sum
0  1    3
1  2    4>
```

jreback replied:
i just want to see a more verbose message that this is not recommended
we could also raise but that's a bit harsh

deniederhut replied:
Okay... how about something like:

> Column name '{key}' collides with a built-in method, which will cause unexpected attribute behavior

jreback replied:
sure

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -17,6 +18,7 @@ class TestABCClasses(object):
     df = pd.DataFrame({'names': ['a', 'b', 'c']}, index=multi_index)
     sparse_series = pd.Series([1, 2, 3]).to_sparse()
     sparse_array = pd.SparseArray(np.random.randn(10))
+    series = pd.Series([1, 2, 3])
```
u can remove this

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,33 @@ def test_abc_types(self):
         assert isinstance(pd.Period('2012', freq='A-DEC'), gt.ABCPeriod)
+
+
+class TestABCWarnings(object):
```
u don't need a class here
u can just do this as a function (more pytest idiomatic)

deniederhut replied:
I didn't see anything in the contributing guide about test style, so I was following the style of the code nearby.

jreback replied:
http://pandas.pydata.org/pandas-docs/stable/contributing.html#test-driven-development-code-writing

pytest section -

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,33 @@ def test_abc_types(self):
+            self.df['three'] = self.df.two + 1
+            assert len(w) == 0
+            assert self.df.three.sum() > self.df.two.sum()
+        with catch_warnings(record=True) as w:
```
put blank lines in between these

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +40,33 @@ def test_abc_types(self):
+            self.df.four = self.df.two + 2
+        with tm.assert_produces_warning(UserWarning):
+            # warn when column has same name as method
+            self.df['sum'] = self.df.two
```
test with self.df.sum as well

deniederhut committed:
CLN: reformatting and cleanup of setting tests

deniederhut committed:
ENH: Clarifies wording of warning message for column name collision

deniederhut committed:
BUG: coerces _set_item key to a string

jreback commented on pandas/core/generic.py:
```diff
@@ -1787,6 +1787,10 @@ def _slice(self, slobj, axis=0, kind=None):
         return result
 
     def _set_item(self, key, value):
+        if callable(getattr(self, str(key), None)):
```
check if its a string first, e.g. 

```
if isinstance(key, compat.text_types) and callable(getattr(self, key, None))):
 ......
```

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +39,37 @@ def test_abc_types(self):
+         'two': pd.Series([1., 2., 3., 4.], index=['a', 'b', 'c', 'd'])}
+    df = pd.DataFrame(d)
+
+    with catch_warnings(record=True) as w:
```
this should NOT show a warning, right?

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +39,37 @@ def test_abc_types(self):
+        assert len(w) == 0
+        assert df.three.sum() > df.two.sum()
+
+    with catch_warnings(record=True) as w:
```
nor this one

jreback commented on pandas/tests/dtypes/test_generic.py:
```diff
@@ -38,3 +39,37 @@ def test_abc_types(self):
+    df = pd.DataFrame(d)
+
+    with catch_warnings(record=True) as w:
+        # successfully add new column
```
add a statement that the first 3 should NOT be showing warnigns (I know you are checking which is fine)

jreback commented:
@deniederhut looks pretty good. rebase on master, some changes and you prob need to force push.

jreback commented:
also needs a whatsnew note; put this in the api-breaking changes section. I *think* we may need a full on example to highlite this.

deniederhut commented:
Wilco. 

And I might be missing something, but we are just issuing warnings, right? This shouldn't break any clients unless they are configured to raise warnings as exceptions?

deniederhut committed:
FIX: only warn getattr when key is str

deniederhut committed:
TST: adds comments about tests which should not warn

jreback commented:
yeah isn't it just a warning then no big deal

however I think we may want to consider making this an error in the future (so can make an issue for that)

deniederhut commented:
Would you like me to add to the whatsnew that these warnings are subject to escalation in future releases?

jreback commented:
no but pls open an issue to re-evaluate in the future

deniederhut committed:
DOC: adds column/attribute warnings to whatsnew

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -57,6 +56,63 @@ using :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedelta`)
+  In[1]: df = pd.DataFrame({'one': [1., 2., 3.]})
+  In[2]: df.two = [4, 5, 6]
+
+which does not raise any obvious exceptions, but also does not create a new column:
```
which -> This

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -57,6 +56,63 @@ using :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedelta`)
+  1  2.0
+  2  3.0
+
+and creating a column whose name collides with a method or attribute already in the instance
```
Creating a column (use sentences if possible here).

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -57,6 +56,63 @@ using :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedelta`)
+.. code-block:: ipython
+
+  In[2]: df.two = [4, 5, 6]
+  UserWarning: Pandas doesn't allow Series to be assigned into nonexistent columns - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
```
Can you put this whole section in indexing.rst (below the attribute access section). Put a reference to it from the whatsnew (you can leave some of it here as well; usually we make the whatsnew a smaller version of what we put in the docs proper).

jreback reviewed (approved):
docs look good! some small changes.

deniederhut committed:
DOC: moves examples from whatsnew to indexing

deniederhut committed:
FIX: replaces ndim check with ABC subclass check

deniederhut commented:
Okay, I _think_ I've gotten the docs correct

deniederhut committed:
CLN: breaks line to avoid 81 char length

deniederhut committed:
TST: renames 'bool' columns in pytables test to avoid collision warning

deniederhut committed:
BUG: fixes 'string_types' encoding failure with unicode cols in py2

As part of warning check, object type of potential attributes was checked
for subtypes of pd.compat.str_types before being checked for overlap with
methods defined on ndframes. This causes decode errors in Python2 when
users attempt to add columns with unicode column names. Fix is to compare
against `str`.

deniederhut committed:
BUG: fixes reversion of isnull/isna produced in merge

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -1906,6 +1906,10 @@ def _slice(self, slobj, axis=0, kind=None):
         return result
 
     def _set_item(self, key, value):
```
Is this a single-point-of-contact that all (most?  many?) setter methods go through?  i.e. will the various `loc.__setitem__` paths eventually wind through here?

gfyoung replied:
Yes, I believe that is correct.

deniederhut replied:
That is my understanding, yes. Here is an example of setting while using `.loc`:

```python
import pandas as pd
df = pd.DataFrame({'one': [0, 1, 2]})
df.loc[:, 'sum'] = df.one.sum()
```

```
UserWarning: Column name 'sum' collides with a built-in method, which will cause unexpected attribute behavior
  self._set_item(key, value)

```

jreback replied:
this point is *only* for ``[]`` setting, e.g. setting a column on a DF or an element on a Series. ``.loc/.iloc`` are handled in core/indexing.py

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +59,51 @@ using the :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedel
+.. _whatsnew_0210.enhancements.column-creation:
+
+Improved warnings when attempting to create columns
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
the underline length must match the title.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +59,51 @@ using the :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedel
+  1  2.0  7.0
+  2  3.0  9.0>
+
+Both of these now raise a ``UserWarning`` about the potential for unexpected behavior. See `Attribute Access <https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access>`__.
```
use `` :ref:`Attribute Access <indexing.attribute_access>` ``

jreback replied:
the section in indexing.rst

jreback commented on pandas/core/generic.py:
```diff
@@ -1906,6 +1906,10 @@ def _slice(self, slobj, axis=0, kind=None):
         return result
 
     def _set_item(self, key, value):
+        if isinstance(key, str) and callable(getattr(self, key, None)):
```
use compat.string_types rather than ``str`` here

deniederhut replied:
Using `string_types` causes failures in Python2 when creating columns that have unicode characters in their names. See c90aa2266137d78f4786b85ff3a5cd08ff0c39fd.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -60,6 +59,51 @@ using the :func:`to_numeric` function (or :func:`to_datetime`, :func:`to_timedel
    df['C'] = pd.to_numeric(df['C'], errors='coerce')
    df.dtypes
 
+.. _whatsnew_0210.enhancements.column-creation:
```
call this ``attribute_access`` (rather than ``column-creation``)

jreback reviewed (changes requested):
some doc comments. lgtm. ping on green for a final look.

deniederhut committed:
DOC: fixes references in indexing and whatsnew

deniederhut commented:
@jreback I think we might be ready for the final check

jreback commented:
thanks @deniederhut nice changes!

jreback commented:
@deniederhut 
see https://travis-ci.org/MacPython/pandas-wheels/jobs/262048673

this is a 32-bit python 2.7 build; there are some warnings at the end. See if you can repro (you can try using a 64-bit python 2.7 build)

jorisvandenbossche commented:
I have some general comments on this PR (sorry for being late with this feedback):

* I like that you now get a warning when trying to create a new column with `df.col_name = [..]` (which silently fails at creating a new column), that's very useful! However, the case when the name you try to set is a builtin, eg `df.mean = [...]` doesn't raise a warning. 
  I think it would be nice to add a warning then as well (as it is actually a more harmful case, it is not only silently failing, but also overwriting your `mean()` method)

* IMO the warning when setting a new column with `df['new_col'] = ..` which has the same name as a builtin method is a bit overkill. I understand it is confusing from time to time that the attribute access does not always work, but IMO there are many more people who use column names like `mean` without any problem that would now get warnings (one example that this is common is applying functions in groupby operations that create such warnings like `s.groupby(..).agg(['mean', 'median'])`). So I am -1 on this change. 
  See also comment above about `df.mean = ..` not raising a warning while `df['mean] = ..` does, but that's something else as it is actually overwriting the method (and this can also be easily fixed to raise I think). 
  Last, if we keep this warning, I think we should also do it for attributes and not only methods? (eg currently `df['shape'] does not raise a warning`

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -3357,6 +3361,12 @@ def __setattr__(self, name, value):
             except (AttributeError, TypeError):
+                if isinstance(self, ABCDataFrame) and (is_list_like(value)):
+                    warnings.warn("Pandas doesn't allow Series to be assigned "
+                                  "into nonexistent columns - see "
```
I would write this a bit more generic, as it does not only raise for Series objects (and I also find it a bit confusing, as pandas certainly allows assigning Series objects, just with another syntax). 

Maybe something like "pandas doesn't allow to add a new column using attribute access" ? (can certainly be improved further)

jorisvandenbossche reviewed (commented):
and also one specific code comment

deniederhut commented:
Happy to dig into this some more - two questions:

1. What is the consensus among the maintainers about warning on name collision?
2. Should this be submitted as a separate PR?

jreback commented:
new PR

shoyer commented:
I also just saw this PR. I agree with @jorisvandenbossche we should not raise warnings when setting column names that conflict with built-in methods. This would be a major source of annoyance for users.

I think we're pretty clear that attribute style access for columns is a convenience feature, not something to be relied on for arbitrary names. In contrast, we *do* guarantee that `__getitem__` works for any valid column name.

jorisvandenbossche commented:
@deniederhut do you want to do a PR for this?

@jreback what do you think about the concerns of @shoyer and me? Fine with changing this part of the PR back?

deniederhut commented:
I'll take a look this weekend

On Wed, Aug 16, 2017, 08:13 Joris Van den Bossche <notifications@github.com>
wrote:

> @deniederhut <https://github.com/deniederhut> do you want to do a PR for
> this?
>
> @jreback <https://github.com/jreback> what do you think about the
> concerns of @shoyer <https://github.com/shoyer> and me? Fine with
> changing this part of the PR back?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/16951#issuecomment-322766464>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGjNrIJP7vd57k2d7VACiSiUdIES21mXks5sYurrgaJpZM4OZEMp>
> .
>

jorisvandenbossche commented:
@deniederhut cool! 
I also just opened https://github.com/pandas-dev/pandas/issues/17268 to keep track of this
